taichi/runtime/gfx/runtime.cpp

```diff
@@ -155,47 +153,39 @@ class HostDeviceContextBlitter {
     TI_ASSERT(device_->map(*device_ret_buffer_, &device_base) ==
               RhiResult::success);
 
-#define TO_HOST(short_type, type, offset)                            \
-  if (dt->is_primitive(PrimitiveTypeID::short_type)) {               \
-    const type d = *(reinterpret_cast<type *>(device_ptr) + offset); \
-    host_result_buffer_[offset] =                                    \
-        taichi_union_cast_with_different_sizes<uint64>(d);           \
-    continue;                                                        \
-  }
-
+    void *ctx_result_buffer = host_ctx_.get_context().result_buffer;
     for (int i = 0; i < ctx_attribs_->rets().size(); ++i) {
-      // Note that we are copying the i-th return value on Metal to the i-th
-      // *arg* on the host context.
       const auto &ret = ctx_attribs_->rets()[i];
       void *device_ptr = (uint8_t *)device_base + ret.offset_in_mem;
+      void *host_ptr = (uint8_t *)ctx_result_buffer + ret.offset_in_mem;
       const auto dt = PrimitiveType::get(ret.dtype);
       const auto num = ret.stride / data_type_size_gfx(dt);
       for (int j = 0; j < num; ++j) {
         // (penguinliong) Again, it's the module loader's responsibility to
         // check the data type availability.
-        TO_HOST(u1, uint1, j)
-        TO_HOST(i8, int8, j)
-        TO_HOST(u8, uint8, j)
-        TO_HOST(i16, int16, j)
-        TO_HOST(u16, uint16, j)
-        TO_HOST(i32, int32, j)
-        TO_HOST(u32, uint32, j)
-        TO_HOST(f32, float32, j)
-        TO_HOST(i64, int64, j)
-        TO_HOST(u64, uint64, j)
-        TO_HOST(f64, float64, j)
-        if (dt->is_primitive(PrimitiveTypeID::f16)) {
-          const float d = fp16_ieee_to_fp32_value(
-              *reinterpret_cast<uint16 *>(device_ptr) + j);
-          host_result_buffer_[j] =
-              taichi_union_cast_with_different_sizes<uint64>(d);
-          continue;
-        }
+#define COPY_TO_HOST_BUFFER(short_type, type, offset)                \
```

Comment:
Can we simply memcpy the whole buffer from device_ptr to host_ptr instead of copying one by one?

Reply:
from device_base to ctx_result_buffer

Reply:
Simplified